### ISSUE TITLE
Tests for SOC dual variable recovery with axis options

### ISSUE THREAD

**Issue by rileyjmurray** [cvxgrp/cvxpy#936]
PR #910 introduced dual variable recovery for general conic constraints. The discussion on that PR mentioned that SOC needs additional testing (specifically with the axis option). Follow-up discussion on a separate issue #933 suggests that the current SOC dual variable recovery isn't respecting argument shapes or axis arguments properly. We need to add tests soon (certainly before a 1.1 release #913). @psa0000 if you can provide an explicit example of failing code that would be helpful.

**Comment by psa0000** [comment]
# To clarify end of #933 and explain test program

[github936_2.zip](https://github.com/cvxgrp/cvxpy/files/4175275/github936_2.zip)

algs=[..] controls test program i.e.

## cvxpy1.0.25
GUROBI issues in variable recovery
e.g.
```
algs=['ECOS','SCS','MOSEK','GUROBI']
```
will crash after GUROBI
['ECOS','SCS','MOSEK']  OK

## cvxpu1.1.0a3

**rileyjmurray**: my  point was that MOSEK and ECOS are now crashing in the same way

GUROBI  issues in variable recovery FIXED
MOSEK, ECOS and possibly others have issue

e.g.
```
algs=[''GUROBI','ECOS',]
```
will crasdh after ECOS or, alternatively, MOSEK

## MOSEK crash
```
.............................
  Primal.  obj: 1.8496515269e-09    nrm: 2e+00    Viol.  con: 5e-10    var: 0e+00    cones: 0e+00
  Dual.    obj: 2.0535543985e-09    nrm: 1e+00    Viol.  con: 0e+00    var: 2e-10    cones: 0e+00
Traceback (most recent call last):
..............................
prob.solve(solver=cvxpy.MOSEK,save_file=fileo+'.mps.gz',verbose=verbose,mosek_params=mosekpar)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 325, in solve
    return solve_func(self, *args, **kwargs)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 642, in _solve
    self.unpack_results(solution, solving_chain, inverse_data)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 882, in unpack_results
    self.unpack(solution)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 840, in unpack
    c.save_dual_value(solution.dual_vars[c.id])
  File "C:\IntelPython3\lib\site-packages\cvxpy\constraints\second_order.py", line 121, in save_dual_value
    cone_size = 1 + self.args[1].shape[self.axis]
IndexError: tuple index out of range
Press any key to continue . . .
```

## or ECOS crash
```
.....................
OPTIMAL (within feastol=1.5e-12, reltol=2.4e+02, abstol=1.3e-10).
Runtime: 0.018958 seconds.

Traceback (most recent call last):
...........................
    prob.solve(solver=cvxpy.ECOS,verbose=verbose)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 325, in solve
    return solve_func(self, *args, **kwargs)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 642, in _solve
    self.unpack_results(solution, solving_chain, inverse_data)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 882, in unpack_results
    self.unpack(solution)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 840, in unpack
    c.save_dual_value(solution.dual_vars[c.id])
  File "C:\IntelPython3\lib\site-packages\cvxpy\constraints\second_order.py", line 121, in save_dual_value
    cone_size = 1 + self.args[1].shape[self.axis]
IndexError: tuple index out of range
```

**Comment by rileyjmurray** [comment]
Thank you for the zip file. I understand that MOSEK and ECOS are crashing in the same way. This is to be expected, because the relevant portion of the code is the same no matter which solver you select here.

The source of the error is almost certainly line 84 of your test file. There you use the axis option to an SOC constraint.

We'll investigate and post a fix hopefully within the week.

**Comment by rileyjmurray** [comment]
Separately, would you mind removing the section of your comment about 1.0.25? The cause of that problem is unrelated to the issue here. As written, the mention of 1.0.25 causes confusion.

(The issue we're investigating here didn't show up in 1.0.25 because 1.0.25 didn't even attempt to recover dual variables to pure conic constraints, SOC and Exp inclusive.)

**Comment by psa0000** [comment]
`rileyjmurray` : I have removed the reference to 1.0.25. But I am still unclear the link between  fixing Gurobi and mucking up the others

**Comment by rileyjmurray** [comment]
@psa0000 I ran your example. When you change line 130 to include ``dual in [0, 1]`` instead of just ``dual in [0]``, then GUROBI fails in the exact same way as ECOS and MOSEK. I'm working on a fix now.

**Comment by rileyjmurray** [comment]
The issue is that in the following code (where ``riska`` is a numpy array of shape (N, m)):
```
    con = []
    s = cvxpy.Variable(N, name='s')
    trade = cvxpy.Variable((N, m))
    con += [cvxpy.SOC(s[i], riska[i] + trade[i], axis=1) for i in range(N)]
```
That code fails in all cases where we attempt dual variable recovery, because the axis is specified in an inappropriate way. Specifically, the expression ``expr = riska[i] + trade[i]`` is of shape ``(m,)``, and has ``expr.shape[1]`` is out of bounds.

Since the input was malformed, it's inevitable that cvxpy would throw an error here. However, we should have thrown an error the moment the SOC constraint object was created. This way the user wouldn't encounter strange behavior such as ``solver='GUROBI', QCPDual=0`` succeeding where other settings fail.

If you run your code without the axis option (or with setting ``axis=0``), then all solvers produce the same result.

**Comment by psa0000** [comment]
Thanks, and my apologies for my axis blunder.  

I have now provisionally tested my real code, which seems fine. Though I did not test Gurobi, as my evaluation has ended.  

I look forward to "the return" of SCS

**Comment by rileyjmurray** [comment]
No worries. It was helpful to look over your example in a bit of detail. For one thing I've identified some input checking SOC needs, and for another thing I've figured out a good test case for the *actual* SOC-axis atom in the process. I will close this issue once those a PR is merged with those changes.

As to SCS, I realize we didn't follow up on that in detail in your other GitHub issue. However I'm confused, because the output you provided in 
https://github.com/cvxgrp/cvxpy/issues/933#issuecomment-581598752 clearly states that you have SCS version ``2.1.1.post2`` installed. Can you try again with SCS? If you experience a program crash then please create a GitHub issue for that (e.g. with a title "Windows: CVXPY install via pip is missing SCS").

**Comment by psa0000** [comment]
I had SCS installed in 1.0.25. But gutted my CVXPY to force it to upgrade to th alpha 1.1.0a3, which came with no SCS.

**Comment by psa0000** [comment]
```
alg= SCS  m= 6
==>  cvxpy_0_SCS_5_6A   DCP= True
Traceback (most recent call last):
  File "C:\Users\psa\Dropbox\GitPython\michal1\github936_3.py", line 141, in <module>
    trade,status=mlsolvecvxD(name1+key,n,m,riska,riskb,dic)
  File "C:\Users\psa\Dropbox\GitPython\michal1\github936_3.py", line 93, in mlsolvecvxD
    status=solvecvx(prob,fileo,dic)
  File "C:\Users\psa\Dropbox\GitPython\michal1\github936_3.py", line 55, in solvecvx
    prob.solve(solver=cvxpy.SCS,verbose=verbose)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 325, in solve
    return solve_func(self, *args, **kwargs)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 639, in _solve
    data, solving_chain, inverse_data = self.get_problem_data(solver, gp)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 425, in get_problem_data
    solving_chain = self._construct_chain(solver=solver, gp=gp)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 549, in _construct_chain
    candidate_solvers = self._find_candidate_solvers(solver=solver, gp=gp)
  File "C:\IntelPython3\lib\site-packages\cvxpy\problems\problem.py", line 489, in _find_candidate_solvers
    raise error.SolverError("The solver %s is not installed." % solver)
cvxpy.error.SolverError: The solver SCS is not installed.
Press any key to continue . . .
```

**Comment by rileyjmurray** [comment]
> I had SCS installed in 1.0.25. But gutted my CVXPY to force it to upgrade to th alpha 1.1.0a3, which came with no SCS.

I see. You have a very unusual python environment at this point. My recommendation would be to create a new environment entirely (either virtualenv, or conda) and then reinstall the pre-release of cvxpy. Windows builds for version 1.1.0a3 are available on pypi and the cvxgrp anaconda channel. If you use the anaconda channel, then you should install SCS *first*, using the conda-forge channel.

Bear in mind that you will need to provide appropriate flags to conda or pip to make sure you get the alpha release 1.1.0a3.
